core/services/gateway/handlers/functions/allowlist/allowlist.go

```diff
@@ -172,17 +204,113 @@ func (a *onchainAllowlist) updateFromContractV1(ctx context.Context, blockNum *b
 	if err != nil {
 		return errors.Wrap(err, "unexpected error during functions_allow_list.NewTermsOfServiceAllowList")
 	}
-	addrList, err := tosContract.GetAllAllowedSenders(&bind.CallOpts{
+
+	var allowedSenderList []common.Address
+	if !a.config.StoreAllowedSendersEnabled {
+		allowedSenderList, err = tosContract.GetAllAllowedSenders(&bind.CallOpts{
+			Pending:     false,
+			BlockNumber: blockNum,
+			Context:     ctx,
+		})
+		if err != nil {
+			return errors.Wrap(err, "error calling GetAllAllowedSenders")
+		}
+	} else {
+		err = a.syncStoredAllowedAndBlockedSenders(ctx, tosContract, blockNum)
+		if err != nil {
+			return errors.Wrap(err, "failed to sync the stored allowed and blocked senders")
+		}
+
+		allowedSenderList, err = a.getAllowedSendersInRange(ctx, tosContract, blockNum)
+		if err != nil {
+			return errors.Wrap(err, "failed to get allowed senders in rage")
+		}
+	}
+
+	a.update(allowedSenderList)
+	return nil
+}
+
+func (a *onchainAllowlist) getAllowedSendersInRange(ctx context.Context, tosContract *functions_allow_list.TermsOfServiceAllowList, blockNum *big.Int) ([]common.Address, error) {
+	allowedSenderList := make([]common.Address, 0)
+	count, err := tosContract.GetAllowedSendersCount(&bind.CallOpts{
 		Pending:     false,
 		BlockNumber: blockNum,
 		Context:     ctx,
 	})
 	if err != nil {
-		return errors.Wrap(err, "error calling GetAllAllowedSenders")
+		return nil, errors.Wrap(err, "unexpected error during functions_allow_list.GetAllowedSendersCount")
+	}
+
+	throttleTicker := time.NewTicker(time.Duration(a.config.FetchingDelayInRangeSec) * time.Second)
+	for idxStart := uint64(0); idxStart < count; idxStart += uint64(a.config.OnchainAllowlistBatchSize) {
+		<-throttleTicker.C
+
+		idxEnd := idxStart + uint64(a.config.OnchainAllowlistBatchSize)
+		if idxEnd >= count {
+			idxEnd = count - 1
```

Comment:
Where is the logic to remove blocked senders from the database?

Reply:
this is done in the `syncStoredAllowedAndBlockedSenders`  we are doing this:
```
                 err = a.syncStoredAllowedAndBlockedSenders(ctx, tosContract, blockNum)
		if err != nil {
			return errors.Wrap(err, "failed to sync the stored allowed and blocked senders")
		}

		allowedSenderList, err = a.getAllowedSendersInRange(ctx, tosContract, blockNum)
		if err != nil {
			return errors.Wrap(err, "failed to get allowed senders in rage")
		}
```

The idea here is that onchain whenever a user is blocked its removed from the allowlist and added to the blocked one. So, if someone is blocked we first run `syncStoredAllowedAndBlockedSenders` that will delete blocked users from the DB and then `getAllowedSendersInRange` that will fetch the fresh list without including any blocked address.